README.md
Chroma-Hash
===========

**A sexy, secure visualization of password field input**

Chroma-Hash is a jQuery plugin that dynamically visualizes secure text-field values using ambient color bars

Password entry can be frustrating, especially with long or difficult passwords. On a webpage, secure fields obscure your input with •'s, so others can't read it. Unfortunately, neither can you—you can't tell if you got your password right until you click "Log In".

Chroma-Hash displays a series of colored bars at the end of field inputs so you can instantly see if your password is right. Chroma-Hash takes an MD5 hash of your input and uses that to compute the colors in the visualization. The resulting color pattern is non-reversible, so no one could know what your password just from the colors.

For a more detailed description of Chroma-Hash, be sure to [check out this blog post about it.](http://mattt.me/2009/chroma-hash-revisited/).

## Demo

Actions speak louder than words. Try it out at: [http://mattt.github.com/Chroma-Hash/](http://mattt.github.com/Chroma-Hash/)

## Usage

```javascript
$("input:password").chromaHash({bars: 3, salt:"7be82b35cb0199120eea35a4507c9acf", minimum:6});
```

- `bars` : number of bars displayed (1,2,3, or 4)
- `salt` : value to be appended when calculating hash function
- `minimum` : minimum number of characters needed for grayscale bars to be displayed in color

## Requirements

- jQuery 1.3+

## Ports, Implementations, and Adaptations

Chroma-Hash is also available in your choice of Javascript libraries and languages:

- [CHChromaHashView (iOS) - @mattt (Mattt Thompson)](https://github.com/mattt/CHChromaHashView)
- [ChromaHashView (Android) - @michaelevans (Michael Evans)](https://github.com/MichaelEvans/ChromaHashView)
- [YUI3 - @foxxtrot (Jeff Craig)](http://github.com/foxxtrot/Chroma-Hash)
- [Prototype - @wki (Wolfgang Kinkeldei)](http://github.com/wki/Chroma-Hash)
- [pyChroma (Python) - @leegao (Lee Gao)](http://github.com/leegao/pyChroma)
- [ChromaHash4j (Java) - @efi (Thomas Efer)](http://github.com/efi/ChromaHash4j)

> For the more acoustically-inclined, be sure to check out [Sonic-Hash](http://mattt.github.com/Sonic-Hash/), which is like its older brother, but uses sounds instead of colors.

## Credit

Inspired by Arc90 Lab's HashMask, by Chris Dary
[http://lab.arc90.com/2009/07/hashmask.php](http://lab.arc90.com/2009/07/09/hashmask-another-more-secure-experiment-in-password-masking/)

Chroma-Hash uses Paul Johnston's Javascript MD5 implementation, and is distributed under the BSD License
See [http://pajhome.org.uk/crypt/md5](http://pajhome.org.uk/crypt/md5) for more info.

Thanks to [Ian Young](http://www.iangreenleaf.com/) for his [suggestion to represent color components in 4-bits](http://blog.iangreenleaf.com/2009/08/making-chroma-hash-less-leaky.html) to increase overall security.

## Contact

Mattt Thompson

- http://github.com/mattt
- http://twitter.com/mattt
- m@mattt.me

## License

Chroma-Hash is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
## Version 1.0.1 [2014-06-25]

- [Bugfix] Fixed rendering issues in Chrome and Firefox

## Version 1.0.0 [2014-06-25]

- [Bugfix] Fixed race condition that caused colors to differ slightly when typing quickly
- [Changed] Using CSS transitions when supported, falling back on jQuery `animate()` (joshdick)
- [Changed] Refactored and modernized code base

## Version 0.4.2 [2013-11-16]

- [Bugfix] Fixed issue where incorrect colors appear when typing quickly

## Version 0.4.1 [2012-8-9]

- [Changed] Renamed chroma-hash.js to jquery.chroma-hash.js, in order to better-communicate jQuery dependency

## Version 0.4.0 [2012-8-9]

- [Bugfix] Chroma-Hash function now works for password elements that do not have an id attribue
- [Changed] Updated and modernized example page to HTML5
- [Changed] Color transitions now use CSS3 transitions, rather than jQuery Color Animations

## Version 0.3.0 [2009-8-12]

- [Changed] jQuery UI Dependency Removed: Chroma-Hash now includes jQuery Color Animations (John Resig), which was the critical component from jQuery UI. (On suggestion of Stan Wiechers, Joel Moss, and others)
- [New] Low Bit Rounding: Each color channel has its least significant digit rounded to zero (eg. 0xF4 -> 0xF0, 0x42 -> 0x4). This increases the number of collisions per color, and thereby improves the overall security of the visualization. Since the differences in colors within a particular 4-byte range are nearly imperceptible to humans, this change does not affect general aesthetics. (Thanks to
[Ian Young](http://blog.iangreenleaf.com/2009/08/making-chroma-hash-less-leaky.html))
- [Bugfix] CSS Positioning Errors: Positioning of color bars is now more robust across different browsers. Successfully tested on Safari 4, Firefox 3.5 and Opera 10

## Version 0.2.0 [2009-8-4]

- [New] Salt: Added parameter for salt to be applied for hash function
- [New] Grayscale: Added parameter to specify minimum number of characters of input needed for colors to appear. Under that length, bars appear in grayscale
- [Bugfix] "Color change effects pause on lost focus" (mikehamer)
- [Changed] Tag-line is now "A sexy, secure visualization of password field input"

## Version 0.1.0 [2009-7-29]

Initial release


chroma-hash.js
/*
 * Chroma-Hash : A sexy, secure visualization of password field input
 * http://github.com/mattt/Chroma-Hash/
 *
 * Copyright (c) 2009-2012 Mattt Thompson (http://mattt.me)
 * Released under the MIT license.
 *
 * Inspired by HashMask by Chris Dary
 * http://lab.arc90.com/2009/07/09/hashmask-another-more-secure-experiment-in-password-masking/
 */

(function($){
  $.fn.extend({
    chromaHash: function(options) {
      var defaults = {
          bars: 3,
          salt: "7be82b35cb0199120eea35a4507c9acf",
          minimum: 6
      };

      var options = $.extend(defaults, options);
      var incrementingID = 0;

      return this.each(function() {
        if (options.bars < 1 || options.bars > 4) {
          console.error("Chroma-Hash expects a number parameter between 1 and 4, given " + options.bars);
          return false;
        }

        var colors = ["primary", "secondary", "tertiary", "quaternary"].slice(0, options.bars);

        var chromaHashesForElement = function(e) {
          var id = $(e).attr('id');

          return $("label.chroma-hash").filter(function(l) {
                      return $(this).attr('for') == id;
                  });
        };

        var trigger = function(e) {
          var input = $(this);

          if (input.val() === "") {
            chromaHashesForElement(this).animate({backgroundColor: "#ffffff", opacity: 0});
            return;
          }

          position = input.position();
          height   = input.outerHeight();
          width    = input.outerWidth();

          chromaHashesForElement(this).each(function(i) {
            properties = {
              "position":   'absolute',
              "opacity":    1.0,
              "left":       position.left + width - 1,
              "top":        position.top,
              "height":     height - 2 + "px",
              "width":      8 + "px",
              "margin":     0 + "px",
              "marginLeft": -8 * (++i) + "px",

              "transition":         "background-color 0.5s",
              "-moz-transition":    "background-color 0.5s",
              "-webkit-transition": "background-color 0.5s",
              "-o-transition":      "background-color 0.5s",
            }

            if ($.browser.safari) {
              properties.marginTop = 3 + "px";
            } else {
              properties.marginTop = 1 + "px";
            }

            $(this).css(properties);
          });

          var id     = input.attr('id');
          var md5    = hex_md5('' + input.val() + ':' + options.salt);
          var colors = md5.match(/([\dABCDEF]{6})/ig);
          $(".chroma-hash").stop();

          if (input.val().length < options.minimum) {
            chromaHashesForElement(this).each(function(i) {
              var g = (parseInt(colors[i], 16) % 0xF).toString(16);
              $(this).css({backgroundColor:"#" + g + g + g});
            });
          } else {
            chromaHashesForElement(this).each(function(i) {
              var color = parseInt(colors[i], 16);
              var red   = (color >> 16) & 255;
              var green = (color >> 8) & 255;
              var blue  = color & 255;
              var hex   = $.map([red, green, blue], function(c, i) {
                return ((c >> 4) * 0x10).toString(16);
              }).join('');

              $(this).css({backgroundColor:"#" + hex});
            });
          }
        };

        $(this).each(function(e) {
          $input = $(this);

          var id = $input.attr('id');
          if (!id || id==="") {
            id = "chroma-hash_" + incrementingID++;
            $input.attr('id', id);
          }

          for(c in colors) {
            $input.after('<label for="' + id + '" class="' + colors[c] + ' chroma-hash"></label>');
          }

          chromaHashesForElement($input).css({backgroundColor: "#FFF", opacity: 0})

          $input.bind('keyup', trigger).bind('blur', trigger);
        });

        /*
         * A JavaScript implementation of the RSA Data Security, Inc. MD5 Message
         * Digest Algorithm, as defined in RFC 1321.
         * Version 2.2 Copyright (C) Paul Johnston 1999 - 2009
         * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
         * Distributed under the BSD License
         * See http://pajhome.org.uk/crypt/md5 for more info.
        */
        var hexcase=0;var b64pad="";function hex_md5(A){return rstr2hex(rstr_md5(str2rstr_utf8(A)));}function b64_md5(A){return rstr2b64(rstr_md5(str2rstr_utf8(A)));}function any_md5(A,B){return rstr2any(rstr_md5(str2rstr_utf8(A)),B);}function hex_hmac_md5(A,B){return rstr2hex(rstr_hmac_md5(str2rstr_utf8(A),str2rstr_utf8(B)));}function b64_hmac_md5(A,B){return rstr2b64(rstr_hmac_md5(str2rstr_utf8(A),str2rstr_utf8(B)));}function any_hmac_md5(A,C,B){return rstr2any(rstr_hmac_md5(str2rstr_utf8(A),str2rstr_utf8(C)),B);}function rstr_md5(A){return binl2rstr(binl_md5(rstr2binl(A),A.length*8));}function rstr_hmac_md5(C,F){var E=rstr2binl(C);if(E.length>16){E=binl_md5(E,C.length*8);}var A=Array(16),D=Array(16);for(var B=0;B<16;B++){A[B]=E[B]^909522486;D[B]=E[B]^1549556828;}var G=binl_md5(A.concat(rstr2binl(F)),512+F.length*8);return binl2rstr(binl_md5(D.concat(G),512+128));}function rstr2hex(C){try{hexcase;}catch(F){hexcase=0;}var E=hexcase?"0123456789ABCDEF":"0123456789abcdef";var B="";var A;for(var D=0;D<C.length;D++){A=C.charCodeAt(D);B+=E.charAt((A>>>4)&15)+E.charAt(A&15);}return B;}function rstr2b64(C){try{b64pad;}catch(G){b64pad="";}var F="ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";var B="";var A=C.length;for(var E=0;E<A;E+=3){var H=(C.charCodeAt(E)<<16)|(E+1<A?C.charCodeAt(E+1)<<8:0)|(E+2<A?C.charCodeAt(E+2):0);for(var D=0;D<4;D++){if(E*8+D*6>C.length*8){B+=b64pad;}else{B+=F.charAt((H>>>6*(3-D))&63);}}}return B;}function rstr2any(K,C){var B=C.length;var J,F,A,L,E;var I=Array(Math.ceil(K.length/2));for(J=0;J<I.length;J++){I[J]=(K.charCodeAt(J*2)<<8)|K.charCodeAt(J*2+1);}var H=Math.ceil(K.length*8/(Math.log(C.length)/Math.log(2)));var G=Array(H);for(F=0;F<H;F++){E=Array();L=0;for(J=0;J<I.length;J++){L=(L<<16)+I[J];A=Math.floor(L/B);L-=A*B;if(E.length>0||A>0){E[E.length]=A;}}G[F]=L;I=E;}var D="";for(J=G.length-1;J>=0;J--){D+=C.charAt(G[J]);}return D;}function str2rstr_utf8(C){var B="";var D=-1;var A,E;while(++D<C.length){A=C.charCodeAt(D);E=D+1<C.length?C.charCodeAt(D+1):0;if(55296<=A&&A<=56319&&56320<=E&&E<=57343){A=65536+((A&1023)<<10)+(E&1023);D++;}if(A<=127){B+=String.fromCharCode(A);}else{if(A<=2047){B+=String.fromCharCode(192|((A>>>6)&31),128|(A&63));}else{if(A<=65535){B+=String.fromCharCode(224|((A>>>12)&15),128|((A>>>6)&63),128|(A&63));}else{if(A<=2097151){B+=String.fromCharCode(240|((A>>>18)&7),128|((A>>>12)&63),128|((A>>>6)&63),128|(A&63));}}}}}return B;}function str2rstr_utf16le(B){var A="";for(var C=0;C<B.length;C++){A+=String.fromCharCode(B.charCodeAt(C)&255,(B.charCodeAt(C)>>>8)&255);}return A;}function str2rstr_utf16be(B){var A="";for(var C=0;C<B.length;C++){A+=String.fromCharCode((B.charCodeAt(C)>>>8)&255,B.charCodeAt(C)&255);}return A;}function rstr2binl(B){var A=Array(B.length>>2);for(var C=0;C<A.length;C++){A[C]=0;}for(var C=0;C<B.length*8;C+=8){A[C>>5]|=(B.charCodeAt(C/8)&255)<<(C%32);}return A;}function binl2rstr(B){var A="";for(var C=0;C<B.length*32;C+=8){A+=String.fromCharCode((B[C>>5]>>>(C%32))&255);}return A;}function binl_md5(K,F){K[F>>5]|=128<<((F)%32);K[(((F+64)>>>9)<<4)+14]=F;var J=1732584193;var I=-271733879;var H=-1732584194;var G=271733878;for(var C=0;C<K.length;C+=16){var E=J;var D=I;var B=H;var A=G;J=md5_ff(J,I,H,G,K[C+0],7,-680876936);G=md5_ff(G,J,I,H,K[C+1],12,-389564586);H=md5_ff(H,G,J,I,K[C+2],17,606105819);I=md5_ff(I,H,G,J,K[C+3],22,-1044525330);J=md5_ff(J,I,H,G,K[C+4],7,-176418897);G=md5_ff(G,J,I,H,K[C+5],12,1200080426);H=md5_ff(H,G,J,I,K[C+6],17,-1473231341);I=md5_ff(I,H,G,J,K[C+7],22,-45705983);J=md5_ff(J,I,H,G,K[C+8],7,1770035416);G=md5_ff(G,J,I,H,K[C+9],12,-1958414417);H=md5_ff(H,G,J,I,K[C+10],17,-42063);I=md5_ff(I,H,G,J,K[C+11],22,-1990404162);J=md5_ff(J,I,H,G,K[C+12],7,1804603682);G=md5_ff(G,J,I,H,K[C+13],12,-40341101);H=md5_ff(H,G,J,I,K[C+14],17,-1502002290);I=md5_ff(I,H,G,J,K[C+15],22,1236535329);J=md5_gg(J,I,H,G,K[C+1],5,-165796510);G=md5_gg(G,J,I,H,K[C+6],9,-1069501632);H=md5_gg(H,G,J,I,K[C+11],14,643717713);I=md5_gg(I,H,G,J,K[C+0],20,-373897302);J=md5_gg(J,I,H,G,K[C+5],5,-701558691);G=md5_gg(G,J,I,H,K[C+10],9,38016083);H=md5_gg(H,G,J,I,K[C+15],14,-660478335);I=md5_gg(I,H,G,J,K[C+4],20,-405537848);J=md5_gg(J,I,H,G,K[C+9],5,568446438);G=md5_gg(G,J,I,H,K[C+14],9,-1019803690);H=md5_gg(H,G,J,I,K[C+3],14,-187363961);I=md5_gg(I,H,G,J,K[C+8],20,1163531501);J=md5_gg(J,I,H,G,K[C+13],5,-1444681467);G=md5_gg(G,J,I,H,K[C+2],9,-51403784);H=md5_gg(H,G,J,I,K[C+7],14,1735328473);I=md5_gg(I,H,G,J,K[C+12],20,-1926607734);J=md5_hh(J,I,H,G,K[C+5],4,-378558);G=md5_hh(G,J,I,H,K[C+8],11,-2022574463);H=md5_hh(H,G,J,I,K[C+11],16,1839030562);I=md5_hh(I,H,G,J,K[C+14],23,-35309556);J=md5_hh(J,I,H,G,K[C+1],4,-1530992060);G=md5_hh(G,J,I,H,K[C+4],11,1272893353);H=md5_hh(H,G,J,I,K[C+7],16,-155497632);I=md5_hh(I,H,G,J,K[C+10],23,-1094730640);J=md5_hh(J,I,H,G,K[C+13],4,681279174);G=md5_hh(G,J,I,H,K[C+0],11,-358537222);H=md5_hh(H,G,J,I,K[C+3],16,-722521979);I=md5_hh(I,H,G,J,K[C+6],23,76029189);J=md5_hh(J,I,H,G,K[C+9],4,-640364487);G=md5_hh(G,J,I,H,K[C+12],11,-421815835);H=md5_hh(H,G,J,I,K[C+15],16,530742520);I=md5_hh(I,H,G,J,K[C+2],23,-995338651);J=md5_ii(J,I,H,G,K[C+0],6,-198630844);G=md5_ii(G,J,I,H,K[C+7],10,1126891415);H=md5_ii(H,G,J,I,K[C+14],15,-1416354905);I=md5_ii(I,H,G,J,K[C+5],21,-57434055);J=md5_ii(J,I,H,G,K[C+12],6,1700485571);G=md5_ii(G,J,I,H,K[C+3],10,-1894986606);H=md5_ii(H,G,J,I,K[C+10],15,-1051523);I=md5_ii(I,H,G,J,K[C+1],21,-2054922799);J=md5_ii(J,I,H,G,K[C+8],6,1873313359);G=md5_ii(G,J,I,H,K[C+15],10,-30611744);H=md5_ii(H,G,J,I,K[C+6],15,-1560198380);I=md5_ii(I,H,G,J,K[C+13],21,1309151649);J=md5_ii(J,I,H,G,K[C+4],6,-145523070);G=md5_ii(G,J,I,H,K[C+11],10,-1120210379);H=md5_ii(H,G,J,I,K[C+2],15,718787259);I=md5_ii(I,H,G,J,K[C+9],21,-343485551);J=safe_add(J,E);I=safe_add(I,D);H=safe_add(H,B);G=safe_add(G,A);}return Array(J,I,H,G);}function md5_cmn(F,C,B,A,E,D){return safe_add(bit_rol(safe_add(safe_add(C,F),safe_add(A,D)),E),B);}function md5_ff(C,B,G,F,A,E,D){return md5_cmn((B&G)|((~B)&F),C,B,A,E,D);}function md5_gg(C,B,G,F,A,E,D){return md5_cmn((B&F)|(G&(~F)),C,B,A,E,D);}function md5_hh(C,B,G,F,A,E,D){return md5_cmn(B^G^F,C,B,A,E,D);}function md5_ii(C,B,G,F,A,E,D){return md5_cmn(G^(B|(~F)),C,B,A,E,D);}function safe_add(A,D){var C=(A&65535)+(D&65535);var B=(A>>16)+(D>>16)+(C>>16);return(B<<16)|(C&65535);}function bit_rol(A,B){return(A<<B)|(A>>>(32-B));}
        });
      }
    });
})(jQuery);


jquery.chroma-hash.js
/*
 * Chroma-Hash : A sexy, secure visualization of password field input
 * http://github.com/mattt/Chroma-Hash/
 *
 * Copyright (c) 2009-2014 Mattt Thompson (http://mattt.me)
 * Released under the MIT license.
 *
 * Inspired by HashMask by Chris Dary
 * http://lab.arc90.com/2009/07/09/hashmask-another-more-secure-experiment-in-password-masking/
 */
(function ($) {
    "use strict";

    var transition = (function () {
        var body = document.body || document.documentElement,
            style = body.style;

        if (typeof style["transition"] === "string") {
            return true;
        }

        var prefixes = ["moz", "webkit", "khtml", "o", "ms"];
        for (var p in prefixes) {
            if (typeof style[prefixes[p] + "Transition"] === "string") {
                return true;
            }
        }

        return false;
    })() ? "css" : "animate";

    $.fn.extend({
        chromaHash: function (options) {
            var defaults = {
                bars: 3,
                salt: "7be82b35cb0199120eea35a4507c9acf",
                minimum: 6
            };
            options = $.extend(defaults, options);

            var incrementingID = 0;

            /*
             * A JavaScript implementation of the RSA Data Security, Inc. MD5 Message
             * Digest Algorithm, as defined in RFC 1321.
             * Version 2.2 Copyright (c) Paul Johnston 1999 - 2009
             * Other contributors: Greg Holt, Andrew Kepert, Ydnar, Lostinet
             * Distributed under the BSD License
             */
            var hexdigest = function (e){function g(e,a,b,c,d,f,n){return h(k(h(h(e,a&b|~a&c),h(d,n)),f),a)}function l(e,a,b,c,d,f,n){return h(k(h(h(e,a&c|b&~c),h(d,n)),f),a)}function m(e,a,b,c,d,f,n){return h(k(h(h(e,a^b^c),h(d,n)),f),a)}function p(e,a,b,c,d,f,n){return h(k(h(h(e,b^(a|~c)),h(d,n)),f),a)}function h(e,a){var b=(e&65535)+(a&65535);return(e>>16)+(a>>16)+(b>>16)<<16|b&65535}function k(e,a){return e<<a|e>>>32-a};for(var a="",b=-1,c,d;++b<e.length;)c=e.charCodeAt(b),d=b+1<e.length?e.charCodeAt(b+1):0,55296<=c&&56319>=c&&56320<=d&&57343>=d&&(c=65536+((c&1023)<<10)+(d&1023),b++),127>=c?a+=String.fromCharCode(c):2047>=c?a+=String.fromCharCode(192|c>>>6&31,128|c&63):65535>=c?a+=String.fromCharCode(224|c>>>12&15,128|c>>>6&63,128|c&63):2097151>=c&&(a+=String.fromCharCode(240|c>>>18&7,128|c>>>12&63,128|c>>>6&63,128|c&63));e=Array(a.length>>2);for(b=0;b<e.length;b++)e[b]=0;for(b=0;b<8*a.length;b+=8)e[b>>5]|=(a.charCodeAt(b/8)&255)<<b%32;a=8*a.length;e[a>>5]|=128<<a%32;e[(a+64>>>9<<4)+14]=a;a=1732584193;b=-271733879;c=-1732584194;d=271733878;for(var f=0;f<e.length;f+=16){var n=a,q=b,r=c,s=d,a=g(a,b,c,d,e[f+0],7,-680876936);d=g(d,a,b,c,e[f+1],12,-389564586);c=g(c,d,a,b,e[f+2],17,606105819);b=g(b,c,d,a,e[f+3],22,-1044525330);a=g(a,b,c,d,e[f+4],7,-176418897);d=g(d,a,b,c,e[f+5],12,1200080426);c=g(c,d,a,b,e[f+6],17,-1473231341);b=g(b,c,d,a,e[f+7],22,-45705983);a=g(a,b,c,d,e[f+8],7,1770035416);d=g(d,a,b,c,e[f+9],12,-1958414417);c=g(c,d,a,b,e[f+10],17,-42063);b=g(b,c,d,a,e[f+11],22,-1990404162);a=g(a,b,c,d,e[f+12],7,1804603682);d=g(d,a,b,c,e[f+13],12,-40341101);c=g(c,d,a,b,e[f+14],17,-1502002290);b=g(b,c,d,a,e[f+15],22,1236535329);a=l(a,b,c,d,e[f+1],5,-165796510);d=l(d,a,b,c,e[f+6],9,-1069501632);c=l(c,d,a,b,e[f+11],14,643717713);b=l(b,c,d,a,e[f+0],20,-373897302);a=l(a,b,c,d,e[f+5],5,-701558691);d=l(d,a,b,c,e[f+10],9,38016083);c=l(c,d,a,b,e[f+15],14,-660478335);b=l(b,c,d,a,e[f+4],20,-405537848);a=l(a,b,c,d,e[f+9],5,568446438);d=l(d,a,b,c,e[f+14],9,-1019803690);c=l(c,d,a,b,e[f+3],14,-187363961);b=l(b,c,d,a,e[f+8],20,1163531501);a=l(a,b,c,d,e[f+13],5,-1444681467);d=l(d,a,b,c,e[f+2],9,-51403784);c=l(c,d,a,b,e[f+7],14,1735328473);b=l(b,c,d,a,e[f+12],20,-1926607734);a=m(a,b,c,d,e[f+5],4,-378558);d=m(d,a,b,c,e[f+8],11,-2022574463);c=m(c,d,a,b,e[f+11],16,1839030562);b=m(b,c,d,a,e[f+14],23,-35309556);a=m(a,b,c,d,e[f+1],4,-1530992060);d=m(d,a,b,c,e[f+4],11,1272893353);c=m(c,d,a,b,e[f+7],16,-155497632);b=m(b,c,d,a,e[f+10],23,-1094730640);a=m(a,b,c,d,e[f+13],4,681279174);d=m(d,a,b,c,e[f+0],11,-358537222);c=m(c,d,a,b,e[f+3],16,-722521979);b=m(b,c,d,a,e[f+6],23,76029189);a=m(a,b,c,d,e[f+9],4,-640364487);d=m(d,a,b,c,e[f+12],11,-421815835);c=m(c,d,a,b,e[f+15],16,530742520);b=m(b,c,d,a,e[f+2],23,-995338651);a=p(a,b,c,d,e[f+0],6,-198630844);d=p(d,a,b,c,e[f+7],10,1126891415);c=p(c,d,a,b,e[f+14],15,-1416354905);b=p(b,c,d,a,e[f+5],21,-57434055);a=p(a,b,c,d,e[f+12],6,1700485571);d=p(d,a,b,c,e[f+3],10,-1894986606);c=p(c,d,a,b,e[f+10],15,-1051523);b=p(b,c,d,a,e[f+1],21,-2054922799);a=p(a,b,c,d,e[f+8],6,1873313359);d=p(d,a,b,c,e[f+15],10,-30611744);c=p(c,d,a,b,e[f+6],15,-1560198380);b=p(b,c,d,a,e[f+13],21,1309151649);a=p(a,b,c,d,e[f+4],6,-145523070);d=p(d,a,b,c,e[f+11],10,-1120210379);c=p(c,d,a,b,e[f+2],15,718787259);b=p(b,c,d,a,e[f+9],21,-343485551);a=h(a,n);b=h(b,q);c=h(c,r);d=h(d,s)}e=[a,b,c,d];a="";for(b=0;b<32*e.length;b+=8)a+=String.fromCharCode(e[b>>5]>>>b%32&255);e=a;a="0123456789ABCDEF";b="";for(d=0;d<e.length;d++)c=e.charCodeAt(d),b+=a.charAt(c>>>4&15)+a.charAt(c&15);return b};

            return this.each(function () {
                if (options.bars < 1 || options.bars > 4) {
                    console.error("Chroma-Hash expects a number parameter between 1 and 4, given " + options.bars);
                    return false;
                }

                var colors = ["primary", "secondary", "tertiary", "quaternary"].slice(0, options.bars);

                var chromaHashesForElement = function (e) {
                    var id = $(e).attr('id');
                    return $("label.chroma-hash").filter(function () {
                        return $(this).attr('for') === id;
                    });
                };

                var trigger = function () {
                    var $input = $(this);

                    if ($input.val() === "") {
                        chromaHashesForElement(this)[transition]({
                            "background": "",
                            "opacity": "0"
                        });

                        return;
                    }

                    var position = $input.position(),
                        height = $input.outerHeight(),
                        width = $input.outerWidth();

                    chromaHashesForElement(this).each(function (i) {
                        var properties = {
                            "position": "absolute",
                            "opacity": "1.0",
                            "left": position.left + width - 1,
                            "top": position.top,
                            "height": height - 2 + "px",
                            "width": "8px",
                            "margin": "0px",
                            "marginLeft": -8 * (++i) + "px",

                            "transition": "background 0.5s",
                            "-moz-transition": "background 0.5s",
                            "-webkit-transition": "background 0.5s",
                            "-khtml-transition": "background 0.5s",
                            "-o-transition": "background 0.5s",
                            "-ms-transition": "background 0.5s",
                        };

                        if (navigator.userAgent.indexOf("Safari") > -1) {
                            properties["marginTop"] = "3px";
                        } else {
                            properties["marginTop"] = "1px";
                        }

                        $(this).css(properties);
                    });

                    var id = $input.attr('id'),
                        digest = hexdigest('' + $input.val() + ':' + options.salt),
                        colors = digest.match(/([\dABCDEF]{6})/ig);

                    if ($input.val().length < options.minimum) {
                        chromaHashesForElement(this).each(function (i) {
                            var gray = (parseInt(colors[i], 16) % 0xF).toString(16);
                            var color = "#" + gray + gray + gray;
                            $(this).stop()[transition]("background", color);
                        });
                    } else {
                        chromaHashesForElement(this).each(function (i) {
                            var color = ("#" + colors[i] + "000000").slice(0, 7);
                            $(this).stop()[transition]("background", color);
                        });
                    }
                };

                $(this).each(function (e) {
                    var $input = $(this);

                    var id = $input.attr('id');
                    if (!id || id === "") {
                        id = "chroma-hash_" + incrementingID++;
                        $input.attr('id', id);
                    }

                    for (var c in colors) {
                        $input.after('<label for="' + id + '" class="' + colors[c] + ' chroma-hash"></label>');
                    }

                    $input.bind('keyup', trigger).bind('blur', trigger);
                });
            });
        }
    });
})(jQuery);
